Submit blog post about tracetesting the OTel demo

adnanrahic opened the pull request:
Hi all!

My colleague @danielbdias wrote this piece based on his recently merged `tracetesting` contribution to the OTel demo that uses [Tracetest](https://github.com/kubeshop/tracetest). I only helped with reviews and formatting. Daniel is out on holiday this week, and he asked me to submit a PR on his behalf.

We thought that it would be nice to blog about it and contribute it to the OTel blog.
Would love to hear your thoughts about it! Do you think it makes sense? 

Thanks!

References:

- PR: https://github.com/open-telemetry/opentelemetry-demo/pull/877
- Original Issue: https://github.com/open-telemetry/opentelemetry-demo/issues/873

---

**Preview**: https://deploy-preview-3053--opentelemetry.netlify.app/blog/2023/testing-otel-demo/

svrnm commented on content/en/blog/2023/testing-otel-demo/index.md:
```diff
@@ -0,0 +1,399 @@
+With contributions from [Adnan Rahić](https://github.com/adnanrahic) and
+[Ken Hamric](https://github.com/kdhamric).
+
+The [OpenTelemetry Demo](../../../ecosystem/demo/) is a system that simulates a
```
```suggestion
The [OpenTelemetry Demo](/ecosystem/demo/) is a system that simulates a
```

svrnm commented on content/en/blog/2023/testing-otel-demo/index.md:
```diff
@@ -0,0 +1,399 @@
+languages, each handling a specific capability of this distributed system. Its
+purpose is to demonstrate how [OpenTelemetry](/) tools and SDKs can be used in
+an application to obtain telemetry for monitoring results and even to
+[track problems](../../../docs/demo/#scenarios) across multiple services.
```
```suggestion
[track problems](/docs/demo/#scenarios) across multiple services.
```

svrnm commented on content/en/blog/2023/testing-otel-demo/index.md:
```diff
@@ -0,0 +1,399 @@
+Trace-based testing is a type of testing that validates a system's behavior by
+triggering an operation against the system and validating its result by checking
+the system output with the emitted
+[traces](../../../docs/concepts/signals/traces/) generated by the system during
```
```suggestion
[traces](/docs/concepts/signals/traces/) generated by the system during
```

svrnm commented on content/en/blog/2023/testing-otel-demo/index.md:
```diff
@@ -0,0 +1,399 @@
+
+![Example of Trace-based test running against a distributed system](./trace-based-testing-diagram.png)
+
+Example of Trace-based test running against a distributed system
```
is this intended to be an image description? I think docsy has some capabilities for that.

adnanrahic replied:
If we cannot have it as an image description, we can just drop this sentence.

adnanrahic replied:
I've removed the image description entirely.

svrnm commented on content/en/blog/2023/testing-otel-demo/index.md:
```diff
@@ -0,0 +1,399 @@
+4. And, completing the checkout process.
+
+We structured two types of tests based on
+[tests that currently exist in the demo](../../../docs/demo/tests/):
```
```suggestion
[tests that currently exist in the demo](/docs/demo/tests/):
```

svrnm commented on content/en/blog/2023/testing-otel-demo/index.md:
```diff
@@ -0,0 +1,399 @@
+We can see two types of assertions:
+
+- The first assertion is against a trace
+  [span](../../../docs/concepts/signals/traces/#spans) that the
```
```suggestion
  [span](/docs/concepts/signals/traces/#spans) that the
```

svrnm commented on content/en/blog/2023/testing-otel-demo/index.md:
```diff
@@ -0,0 +1,399 @@
+[Cypress](https://www.cypress.io/). We call the services through the API used by
+the front end and check if the service interaction between them is correct. We
+also verify if a trace is
+[propagated](../../../docs/concepts/signals/traces/#context-propagation)
```
```suggestion
[propagated](/docs/concepts/signals/traces/#context-propagation)
```

svrnm commented on content/en/blog/2023/testing-otel-demo/index.md:
```diff
@@ -0,0 +1,399 @@
+    assertions:
+    - attr:response.status = 200
+  - selector: span[name="oteldemo.CheckoutService/PlaceOrder" rpc.system="grpc"
+			rpc.method="PlaceOrder" rpc.service="oteldemo.CheckoutService"]
```
```suggestion
            rpc.method="PlaceOrder" rpc.service="oteldemo.CheckoutService"]
```

svrnm commented on content/en/blog/2023/testing-otel-demo/index.md:
```diff
@@ -0,0 +1,399 @@
+    - attr:app.user.id = "2491f868-88f1-4345-8836-d5d8511a9f83"
+    - attr:app.order.items.count = 1
+  - selector: span[name="oteldemo.PaymentService/Charge" rpc.system="grpc"
+			rpc.method="Charge" rpc.service="oteldemo.PaymentService"]
```
```suggestion
            rpc.method="Charge" rpc.service="oteldemo.PaymentService"]
```

svrnm commented on content/en/blog/2023/testing-otel-demo/index.md:
```diff
@@ -0,0 +1,399 @@
+    - attr:rpc.grpc.status_code  =  0
+    - attr:selected_spans.count >= 1
+  - selector: span[name="oteldemo.ShippingService/ShipOrder" rpc.system="grpc"
+			rpc.method="ShipOrder" rpc.service="oteldemo.ShippingService"]
```
```suggestion
            rpc.method="ShipOrder" rpc.service="oteldemo.ShippingService"]
```

svrnm commented on content/en/blog/2023/testing-otel-demo/index.md:
```diff
@@ -0,0 +1,399 @@
+    - attr:rpc.grpc.status_code = 0
+    - attr:selected_spans.count >= 1
+  - selector: span[name="oteldemo.CartService/EmptyCart" rpc.system="grpc"
+			rpc.method="EmptyCart" rpc.service="oteldemo.CartService"]
```
```suggestion
            rpc.method="EmptyCart" rpc.service="oteldemo.CartService"]
```

svrnm commented on content/en/blog/2023/testing-otel-demo/index.md:
```diff
@@ -0,0 +1,399 @@
+test file run in the transaction and also with the "Checkout” step described
+above:
+
+![results from the trace-based test](./trace-based-test-result.png)
```
Since this is text only I recommend to make use of the text directly

adnanrahic replied:
Agreed. Changing to text.

svrnm commented on content/en/blog/2023/testing-otel-demo/index.md:
```diff
@@ -0,0 +1,399 @@
+    "shippingCost": {
+      "currencyCode": "USD",
+    },
+	  // ...
```
```suggestion
        // ...
```

svrnm commented on content/en/blog/2023/testing-otel-demo/index.md:
```diff
@@ -0,0 +1,399 @@
+By doing that, we have a successful call to the service, and it evaluates
+correctly as seen here:
+
+![email-success-formatted.png](./email-success-formatted.png)
```
Is there any reason why tracetest is used in this last screenshot and not jaeger or a console output as before? Also, the picture is really hard to parse on a small screen

adnanrahic replied:
The reasoning is to show the test specs and that the test is passing. I can try re-taking the screenshot and make it bigger.

svrnm commented on content/en/blog/2023/testing-otel-demo/index.md:
```diff
@@ -0,0 +1,399 @@
+on [the issue](https://github.com/open-telemetry/opentelemetry-demo/issues/873)
+as soon as it was opened two months ago.
+
+Daniel Dias is working on Tracetest, an open source tool that allows you to
+develop and test your distributed system with OpenTelemetry. It works with any
+OTel-compatible system and enables trace–based tests to be created. Check it out
+at [https://github.com/kubeshop/tracetest](https://github.com/kubeshop/tracetest).
```
```suggestion
```

svrnm commented:
thanks @adnanrahic , please see the comments I have provided

adnanrahic commented:
Thanks @svrnm! I've applied your suggestions.

I'll re-take the final screenshot to make the screen bigger and more readable.

Is there anything else I missed?

svrnm commented on content/en/blog/2023/testing-otel-demo/index.md:
```diff
@@ -0,0 +1,416 @@
+spec:
+  name: 'Frontend Service'
+  description: Run all Frontend tests enabled in sequence,
+		simulating a process of a user purchasing products on the Astronomy store
```
```suggestion
        simulating a process of a user purchasing products on the Astronomy store
```

adnanrahic commented:
Hi @svrnm 👋 

I've addressed all suggestions and made sure to properly format the post.
I'm happy to make additional edits if needed.

Thanks. 👌

svrnm commented:
@adnanrahic thanks! I'll need a few days to be able to do another round of reviews, apologies, I'll follow up with you as soon as I can.

cartermp reviewed (approved):
Overall I think the blog is in good shape.

chalin reviewed (commented):
I'll make a copyedit pass if I can soon too (though probably after @svrnm :)).

svrnm commented on content/en/blog/2023/testing-otel-demo/index.md:
```diff
@@ -0,0 +1,421 @@
+generates a trace, following these steps:
+
+1. Trigger an operation against the system and collect its output and trace id
+   generated from the operation;
```
```suggestion
   generated from the operation.
```

svrnm reviewed (commented):
minor issues, left, overall lgtm

Let's publish this by the end of the week/early next week as we just had another blog post yesterday

adnanrahic commented on content/en/blog/2023/testing-otel-demo/index.md:
```diff
@@ -0,0 +1,421 @@
+---
+title: Testing the OpenTelemetry Demo with Trace-based Tests
```
Sounds a lot better actually! 👌

adnanrahic commented:
Thanks everyone!

I've applied your last edits. It looks great!

Would be awesome to see it published this week. 🙌

svrnm commented on content/en/blog/2023/testing-otel-demo/index.md:
```diff
@@ -0,0 +1,421 @@
+---
+title: Trace-based Testing the OpenTelemetry Demo
+linkTitle: Testing the OTel Demo
+date: 2023-07-18
```
```suggestion
date: 2023-07-27
```

svrnm commented:
Let's publish it on Thursday then

adnanrahic commented:
Wonderful @svrnm ! 🙌 
Thanks everyone for the reviews. 🙇

adnanrahic commented:
Hey @svrnm ! Just checking in to ask when you plan on merging this? Wanted to know so I can help promote it once it's published. 😄

chalin commented:
Hi @adnanrahic - I've just rebased. Can you pull the rebased PR and then run `npm run fix:dict` and submit the changes? That'll normalize your local cSpell words. Thanks!

svrnm commented:
published! https://opentelemetry.io/blog/2023/testing-otel-demo/

adnanrahic commented:
Thanks a lot, everyone! 🙌
